### ISSUE TITLE
SkipSelf annotation to get parent component Injector no longer working in Ivy (Angular 9 rc.9)

### ISSUE THREAD

**Issue by elca-lan** [angular/angular#34819]
# 🐞 bug report

### Affected Package
The issue is caused by package @angular/core

### Is this a regression?
Yes, with Angular 8 this was still possible.

### Description
I have a component B that is a child of component A. With previous Angular versions I was able to inject the `Injector` of component A into component B by simply using the `@SkipSelf` annotation on the injected service. With that, I could then resolve services in the context of the parent component.

In my use case, both A and B provide the same injection token with different fixed values and component B needs to know what this value was for its parent component.

## 🔬 Minimal Reproduction
https://stackblitz.com/edit/angular-parent-injector-gx5pha

The example on StackBlitz is of course in Angular 8 (since I can't seem to do Angular 9 rc projects) and works as expected. The output on the console is:

<pre><code>
this is "HELLO COMPONENT" and parent has name "THIS IS ROOT"
this is "HELLO COMPONENT" and parent has name "THIS IS ROOT"
</code></pre>

When I download the example and upgrade it to Angular 9 rc.9 I get different output:

<pre><code>
this is "HELLO COMPONENT" and parent has name "THIS IS ROOT"
this is "HELLO COMPONENT" and parent has name "HELLO COMPONENT"
</code></pre>

The second line is produced with the parent `Injector` injection approach.

## 🌍  Your Environment

**Angular Version:**
I tested this with Angular 9 rc.7, rc.8 and rc.9 and the example fails in all of them.

**Anything else relevant?**
In the past I have tried other aproaches to solve this issue, one of them was to use the `Injector.get` method and specifying the `InjectFlags` argument as documented in the [Injector](https://angular.io/api/core/Injector). It seems that in all versions of Angular (that I tested), the injector instance that I get when requesting an `Injector` on a component simply _ignores_ the inject flags - in the source code I can see that the method simply does not use this argument. Somehow I find this very weird since the documentation does not seem to mention that.

**Comment by JoostK** [comment]
I suspect this is caused by this logic:

https://github.com/angular/angular/blob/1a0ee18d62b5bbe1050ed257abd7707899af9514/packages/core/src/render3/di.ts#L349-L352

and just above it also seems to ignore the flags:

https://github.com/angular/angular/blob/1a0ee18d62b5bbe1050ed257abd7707899af9514/packages/core/src/render3/di.ts#L337-L347

**Comment by elca-lan** [comment]
Good to know that my hunch wasn't too far off. The whole `lView` and `tValue` thing was "a bit "confusing...

Would you say that my use case is something that _should_ work - which would make this an actual bug?

**Comment by JoostK** [comment]
@elca-lan looks like a bug to me 👍

**Comment by AndrewKushnir** [comment]
Hi @elca-lan,

As @JoostK pointed out, it looks like it's a bug in Ivy and I confirmed that with a test case that works with View Engine, but fails with Ivy:

```
    it('should support @SkipSelf when injecting Injectors', () => {
      @Component({
        selector: 'parent',
        template: '<child></child>',
        providers: [{
          provide: 'token',
          useValue: 'PARENT',
        }]
      })
      class ParentComponent {
      }

      @Component({
        selector: 'child',
        template: '...',
        providers: [{
          provide: 'token',
          useValue: 'CHILD',
        }]
      })
      class ChildComponent {
        constructor(@SkipSelf() public parentInjector: Injector) {}
      }

      TestBed.configureTestingModule({declarations: [ParentComponent, ChildComponent]});
      const fixture = TestBed.createComponent(ParentComponent);
      fixture.detectChanges();

      const child = fixture.debugElement.query(By.directive(ChildComponent));
      expect(child.componentInstance.parentInjector.get('token')).toBe('PARENT');
    });
```

We have a [PR in progress](https://github.com/angular/angular/pull/34554) that fixes similar issue for `ViewContainerRef`, `ElementRef` and `ChangeDetectorRef` tokens (when they are used with @SkipSelf), but this issue would require a separate fix.

**Comment by Shijir** [comment]
This looks like a duplicate of https://github.com/angular/angular/issues/34066.

**Comment by AndrewKushnir** [comment]
@Shijir, you are right, these problems look similar from API standpoint (in both cases @SkipSelf is not working as expected). However there are different code paths for `ElementRef`/`ViewContainerRef`/`ChangeDetectorRef`s and `Injector`s. We have a [PR in progress](https://github.com/angular/angular/pull/34554) to address the first case and I'm working on a fix with a separate set of tests for the `Injector` case. Thank you.

**Comment by Shijir** [comment]
@AndrewKushnir sounds good. Thanks for clarifying.

**Comment by mehmet-erim** [comment]
Is there any progress?

**Comment by AndrewKushnir** [comment]
Hi @mehmet-erim, while working on this problem we came across several cases that would require additional refactoring to support them. We also found some inconsistencies with the way these cases are handled in View Engine, so we'll need a bit more time to come up with proper solution for Ivy. We'll share more updates once we have them. Thank you.

**Comment by FDIM** [comment]
Nice to find out that my suspicion was right as well. I'll be going with a workaround and passing parentCdRef as an input to component until this is resolved

**Comment by AndrewKushnir** [comment]
FYI, this issue is resolved by #39464 and the fix will be available in v11.0.0. Thank you.
